p2p/transport_test.go

```diff
@@ -0,0 +1,1016 @@
+		func(_ map[ID]Peer, _ Peer) error {
+			time.Sleep(10 * time.Millisecond)
+			return nil
+		},
+	)(mt)
+
+	addr, err := NewNetAddressStringWithOptionalID("127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := mt.Listen(*addr); err != nil {
+		t.Fatal(err)
+	}
+
+	errc := make(chan error)
+
+	go func() {
+		var (
+			pv     = ed25519.GenPrivKey()
+			dialer = NewMultiplexTransport(
+				NodeInfo{
+					ID:         PubKeyToID(pv.PubKey()),
+					ListenAddr: "127.0.0.1:0",
+					Moniker:    "dialer",
+					Version:    "1.0.0",
+				},
+				NodeKey{
+					PrivKey: pv,
+				},
+			)
+		)
+
+		addr, err := NewNetAddressStringWithOptionalID(mt.listener.Addr().String())
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		_, err = dialer.Dial(*addr, peerConfig{})
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		t.Errorf("connection failed: %v", err)
+	}
+
+	_, err = mt.Accept(peerConfig{})
+	if errors.Cause(err) != ErrTransportFilterTimeout {
+		t.Errorf("expected ErrTransportFilterTimeout")
+	}
+}
+
+func TestTransportMultiplexPeerFilterDuplicate(t *testing.T) {
+	var (
+		pv = ed25519.GenPrivKey()
+		mt = NewMultiplexTransport(
+			NodeInfo{
+				ID:         PubKeyToID(pv.PubKey()),
+				ListenAddr: "127.0.0.1:0",
+				Moniker:    "transport",
+				Version:    "1.0.0",
+			},
+			NodeKey{
+				PrivKey: pv,
+			},
+		)
+	)
+
+	addr, err := NewNetAddressStringWithOptionalID("127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := mt.Listen(*addr); err != nil {
+		t.Fatal(err)
+	}
+
+	errc := make(chan error)
+
+	go func() {
+		var (
+			pv     = ed25519.GenPrivKey()
+			dialer = NewMultiplexTransport(
+				NodeInfo{
+					ID:         PubKeyToID(pv.PubKey()),
+					ListenAddr: "127.0.0.1:0",
+					Moniker:    "dialer",
+					Version:    "1.0.0",
+				},
+				NodeKey{
+					PrivKey: pv,
+				},
+			)
+		)
+
+		addr, err := NewNetAddressStringWithOptionalID(mt.listener.Addr().String())
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		_, err = dialer.Dial(*addr, peerConfig{})
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		duplicate := NewMultiplexTransport(
+			NodeInfo{
+				ID:         PubKeyToID(pv.PubKey()),
+				ListenAddr: "127.0.0.1:0",
+				Moniker:    "dialer",
+				Version:    "1.0.0",
+			},
+			NodeKey{
+				PrivKey: pv,
+			},
+		)
+
+		_, err = duplicate.Dial(*addr, peerConfig{})
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		t.Errorf("connection failed: %v", err)
+	}
+
+	if _, err := mt.Accept(peerConfig{}); err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = mt.Accept(peerConfig{})
+	if err, ok := err.(*ErrRejected); ok {
+		if !err.IsDuplicate() {
+			t.Errorf("expected peer to be duplicate")
+		}
+	} else {
+		t.Errorf("expected ErrRejected")
+	}
+}
+
+func TestTransportMultiplexAcceptMultiple(t *testing.T) {
+	var (
+		pv = ed25519.GenPrivKey()
+		mt = NewMultiplexTransport(
```

Comment:
Which part? I'm hesitant as we have a lot of obfuscation in our tests in the project which doesn't necessarily help. I prefer a very verbose order of execution so the test is self-contained. @ebuchman pointed at a good resource regarding the [wrong abstraction](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction) and I think this could be one of those, unless we have very strong arguments what the gains would be, besides a couple of lines spared.

Reply:
You're creating the same transport, so I thought it can be extracted into `createMultiplexTransport` func.

```
func createMultiplexTransport(listenAddr string) {
pv = ed25519.GenPrivKey()
		mt = NewMultiplexTransport(
			NodeInfo{
				ID:         PubKeyToID(pv.PubKey()),
				ListenAddr: "127.0.0.1:0",
				Moniker:    "transport",
				Version:    "1.0.0",
			},
			NodeKey{
				PrivKey: pv,
			},
		)
}
```

Reply:
We can sure do that. My point is then in order to understand any given test-case you have to consult the implementation of `createMultiplexTransport`. Is the main motivation here to reduce repetition?

Reply:
> Is the main motivation here to reduce repetition?

yes

> My point is then in order to understand any given test-case you have to consult the implementation of createMultiplexTransport.

I understand, but that can be said about anything in programming. And that does not mean Tendermint should be 1 function 50000 lines long 😆